### PR TITLE
Refactor JSON deserialization to assist code generation

### PR DESCRIPTION
**Pull request by jdisanti**
More work towards #161. These changes to the rust-runtime facilitate the JSON parser generator by making it possible to create custom errors, and also to expect certain value types or skip entire values in the token stream.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: rust-runtime/smithy-json/src/deserialize.rs [hunk 1]
```diff
@@ -285,20 +211,28 @@ impl<'a> JsonTokenIterator<'a> {
...
             }
             b'f' => {
                 self.expect_literal(b"false")?;
-                Ok(Token::ValueBool(false))
+                Ok(Token::ValueBool {
+                    offset,
+                    value: false,
+                })
             }
             _ => unreachable!(),
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/deserialize.rs:237]: why is this unreachable? it seems like you would hit this if you called expect_bool but there wasn't a bool?

File: rust-runtime/smithy-json/src/deserialize/token.rs [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+    /// Unescapes the string and returns it.
+    /// If the string doesn't need unescaping, it will be returned directly.
+    pub fn to_unescaped(&self) -> Result<Cow<'a, str>, EscapeError> {
+        unescape_string(self.0)
+    }
+}
+
+/// Represents the location of a token
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+pub struct TokenOffset(pub usize);
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/deserialize/token.rs:37]: I would probably just name this `Offset`. We're already in a module named `Token`

File: rust-runtime/smithy-json/src/deserialize/token.rs [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+            }
+        }
+    };
+}
+
+expect_fn!(expect_start_object, StartObject);
+expect_fn!(expect_start_array, StartArray);
+
+/// Expects a string or null token, and if its a string, returns its unescaped value.
+pub fn expect_string_or_null(
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/deserialize/token.rs:125]: I was a little surprised to find no `expect_string`.

File: rust-runtime/smithy-json/src/deserialize/token.rs [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+                )),
+            }
+        }
+    };
+}
+
+expect_fn!(expect_start_object, StartObject);
+expect_fn!(expect_start_array, StartArray);
+
+/// Expects a string or null token, and if its a string, returns its unescaped value.
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/deserialize/token.rs:124]:
```suggestion
/// Expects a string or null token. If the value was a string, its **unescaped** value will be returned
```

File: rust-runtime/smithy-json/src/deserialize/token.rs [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+pub fn expect_string_or_null(
+    token: Option<Result<Token<'_>, Error>>,
+) -> Result<Option<String>, Error> {
+    match token.transpose()? {
+        Some(Token::ValueNull { .. }) => Ok(None),
+        Some(Token::ValueString { value, .. }) => Ok(Some(value.to_unescaped()?.to_string())),
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/deserialize/token.rs:130]: I guess you know you're definitely going to need a string at this point so there is no point in keeping at a `Cow`?

File: rust-runtime/smithy-json/src/deserialize/token.rs [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+        Some(Ok(Token::ValueString {
+            offset: TokenOffset(offset),
+            value: EscapedStr::new(string),
+        }))
+    }
+
+    #[test]
+    fn skip_simple_value() {
+        let mut tokens = json_token_iter(b"null true");
+        skip_value(&mut tokens).unwrap();
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/deserialize/token.rs:251]: so it seems like json_token_iter accepts invalid JSON? not necessarily a problem, just worth documenting what is accepted and what is not

**Review by rcoh — APPROVED** [review]

File: rust-runtime/smithy-json/src/deserialize.rs (hunk 1 above)
Comment (rcoh) [inline rust-runtime/smithy-json/src/deserialize.rs:237]: probably worth commenting why this is unreachable or putting a message in `unreachable`
```suggestion
            _ => unreachable!("this function MUST only be called when the next byte is t or f"),
```

File: rust-runtime/smithy-json/src/deserialize/token.rs (hunk 2 above)
Comment (jdisanti) [reply rust-runtime/smithy-json/src/deserialize/token.rs:37]: Makes sense.

File: rust-runtime/smithy-json/src/deserialize/token.rs (hunk 3 above)
Comment (jdisanti) [reply rust-runtime/smithy-json/src/deserialize/token.rs:125]: I need to think carefully about the null handling. Right now, it just passes off the Option to the builder's set function, and later, if that field was actually required, the fallible builder will fail. I'm not certain if this works correctly for, say, a primitive int field.

File: rust-runtime/smithy-json/src/deserialize/token.rs (hunk 5 above)
Comment (jdisanti) [reply rust-runtime/smithy-json/src/deserialize/token.rs:130]: Exactly. The only time this gets called is when we're moving a string value into a builder.

File: rust-runtime/smithy-json/src/deserialize/token.rs (hunk 6 above)
Comment (jdisanti) [reply rust-runtime/smithy-json/src/deserialize/token.rs:251]: I don't think that's necessarily invalid, but rather, multiple values. I could make it only read one value and then fail if there's more, but that's effectively what the codegen doing with it.

File: rust-runtime/smithy-json/src/deserialize.rs [hunk 7]
```diff
@@ -9,11 +9,25 @@ mod error;
...
 use ErrorReason::*;
 
-/// Returns an Iterator of `Result<Token, Error>` over an slice of bytes.
+/// ### JSON token parser as a Rust iterator
+///
+/// This parser will parse and yield exactly one [Token] per iterator `next()` call.
+/// Validation is done on the fly, so it is possible for it to parse an invalid JSON document
+/// until it gets to the first [Error].
+///
+/// JSON string values are left unescaped in the [Token::ValueString] as an [EscapedStr],
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply rust-runtime/smithy-json/src/deserialize.rs:22]: Yeah. Good catch!

**Review by rcoh — APPROVED** [review]
moving base64 & other refactorings LGTM!

**Pull request merged**